devito/passes/iet/mpi.py

```diff
@@ -45,6 +45,13 @@ def _drop_halospots(iet):
             if f in hs.fmapper and all(i.is_reduction for i in v):
                 mapper[hs].add(f)
 
+    # If a HaloSpot is outside any iteration it is not needed
```

Comment:
Comment is a bit misleading. It is not droping any it is dropping HaloSpot that have an iter (loc_indices) but go lifted out of it meaning there is no need for it, it still kepps halospot that are out of iteration that don't have loc_indices, i.e initial ones (there is one of the test that would fail if it wasn't)

Reply:
So
- Without this pass `test_mpi.py::test_injection_wodup_wtime` fails because you end up with an HaloSpot for a write-to only function outside the time loop (that HaloSpot is not there in master but appears here due to the revamp of dimensions/... of sparse inject). So it actually correctly removes it
- If removed but does not check that loc_indices isn't empty then `test_hoist_haloupdate_from_innerloop` fails as it wrongly removes the read halospot with no loc_indices (Function)  for `f`

Reply:
> write-to only function

What function is that? 
Why do we have a write-only function? is it just a feature of the test?

If the symptom is the presence of a HaloSpot for a write-only function, then the actual issue, I think, is that that HaloSpot should have never been generated in the first place. So the actual problem might be earlier on in the compilation pipeline, in particular here:

https://github.com/devitocodes/devito/blob/master/devito/ir/clusters/algorithms.py#L374

If this were correct, then all we have to do at that point is avoiding to add stuff to `points` if `a.function` doesn't appear in `scope.reads`.

Before I develop any further this concept (if necessary), let's see if we agree on the first part, namely:

> that HaloSpot should have never been generated in the first place

Reply:
So had a quick look, it seems to be a bit more involved bedause some how it has the wrong properties by the time it gets there. i.e you have
```
r0 = ....
f[ ...] += r0
```

but it think it's a read only (is_write False and is_read True) for `f` I'll see if can find the source of the issue

Reply:
copy-pasting from slack for future reference:

I tried dropping [this entire piece of code ](https://github.com/devitocodes/devito/pull/2128/files#diff-e2c8e1150ab045d6300aee639c2af8a7deba5ef56290603d94a6bba2895cc83bR49-R56)and I’ve just got a full pass on test_mpi.py
this seems to contradict what you were saying [here](https://github.com/devitocodes/devito/pull/2128/files#r1273671513) —  in particular, the fact that:
If removed but does not check that loc_indices isn't empty then test_hoist_haloupdate_from_innerloop fails as it wrongly removes the read halospot with no loc_indices (Function) for f